### PR TITLE
Assignment3

### PR DESCRIPTION
**Pull request by TechGlue**
Initial PR
Partner - Nicholas Johnson @NicholasJohnsonDev

### CODE REVIEW

**Comment by GavenElder** [comment]
I will PR this.

**Comment by GavenElder** [comment]
For this assignment do the following:

    Create two interfaces. One to represent the JokeService and one interface to represent outputting the joke to the screen (Console.WriteLine). ✔
    The JokeService will need to have the interface applied to it. ✔
    Create an implementation of the output interface that writes out the joke to the console. ✔
    Implement the Jester class. It should take in both interfaces as dependencies. These dependencies should be null checked. ✔
    The Jester class TellJoke() method should retrieve a joke from the JokeService. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ✔
    Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ✔
    Enable nullability in the .csproj files and ensure no errors or warnings: ✔

Fundamentals

    Ensure you enable:
        nullable reference types is enabled ✔
        net6 targetted ✔
        C# 10.0 ✔
        and enabled .NET analyzers for both projects ✔
    For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
    All of the above should be unit tested ✔
    Choose simplicity over complexity ✔

Extra Credit

    Unit test your implementation that writes the joke out to the screen. How hard could it be to unit test a single line method ;)?
    The Geek jokes API that is being used can also return jokes in a JSON format. Update the JokeService to retrieve jokes using JSON. GetJoke should still return a string. ❌

**Comment by GavenElder** [comment]
Just realized that this is done with a partner. LOL

**Comment by TechGlue** [comment]
I still do appreciate the review 🥸

File: CanHazFunny/CanHazFunny/JokeService.cs [hunk 1]
```diff
@@ -1,15 +1,19 @@
-﻿using System.Net.Http;
+﻿namespace CanHazFunny;
 
-namespace CanHazFunny
+public class JokeService : IJokeService, IJokesOut
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/JokeService.cs:3]:
```suggestion
public class JokeService : IJokeService
```
This should be implemented in a separate class

File: CanHazFunny/CanHazFunny/IJokesOut.cs [hunk 2]
```diff
@@ -0,0 +1,5 @@
+namespace CanHazFunny;
+public interface IJokesOut
+{
+    public void PrintJoke(string? joke);
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/IJokesOut.cs:4]:
```suggestion
    public void PrintJoke(string joke);
```
This probably doesn't need to be nullable.

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs [hunk 3]
```diff
@@ -1,10 +1,43 @@
...
+            .Returns("chuck norris Joke")
+            .Returns("CHUCk NoRRis Joke")
+            .Returns("Joke");
+
+        //Act
+        Jester jester = new(jokeServiceMock.Object, jokeService);
+        jester.TellJoke();
+
+        //Asserts
+        Assert.AreEqual<string>("Joke", stringWriter.ToString().Trim());
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny.Tests/JesterTests.cs:37]: 👍🏻

File: CanHazFunny/CanHazFunny/JokeService.cs [hunk 4]
```diff
@@ -5,8 +5,10 @@ public class JokeService : IJokeService, IJokesOut
     private HttpClient HttpClient { get; } = new();
     public string GetJoke()
     {
-        Uri uri = new("https://geek-jokes.sameerkumar.website/api");
+        //extra credit
+        Uri uri = new("https://geek-jokes.sameerkumar.website/api?format=json");
         string joke = HttpClient.GetStringAsync(uri).Result;
+        joke = System.Text.Json.JsonDocument.Parse(joke).RootElement.GetProperty("joke").ToString();
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/JokeService.cs:11]: Here is some documentation on how to do this by deserializing the JSON into a C# object.
https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-how-to?pivots=dotnet-6-0#how-to-read-json-as-net-objects-deserialize

**Pull request closed without merging**